### PR TITLE
Improve the msg retry mechanism to avoid blocking msgs sending

### PR DESCRIPTION
**Pull request by HJianBo**

### CODE REVIEW

File: src/emqtt.erl [hunk 1]
```diff
@@ -1036,31 +1079,56 @@ should_ping(ConnMod, Sock) ->
...
+        {ok, {publish, #mqtt_msg{packet_id = PacketId}, _Meta}} ->
             ok = eval_msg_handler(State, puback, #{packet_id   => PacketId,
                                                    reason_code => ReasonCode,
                                                    properties  => Properties}),
-            State#state{inflight = maps:remove(PacketId, Inflight)};
+            Inflight1 = maps:remove(PacketId, Inflight),
+            set_inflight_cnt(maps:size(Inflight1)),
+            State#state{inflight = Inflight1};
         error ->
             ?LOG(warning, "Unexpected PUBACK: ~p", [PacketId], State),
```
Comment (zmstone) [inline src/emqtt.erl:1119]: maybe lower to info.

File: src/emqtt.erl [hunk 2]
```diff
@@ -977,7 +1020,7 @@ handle_event(info, {TcpOrSsL, _Sock, Data}, _StateName, State)
 handle_event(info, {Error, _Sock, Reason}, _StateName, State)
     when Error =:= tcp_error; Error =:= ssl_error ->
     ?LOG(error, "The connection error occured ~p, reason:~p",
```
Comment (zmstone) [inline src/emqtt.erl:1022]:
```suggestion
    ?LOG(error, "Connection error ~p, reason:~p",
```

File: src/emqtt.erl [hunk 3]
```diff
@@ -1073,6 +1141,53 @@ delete_inflight_when_full(Packet, State) ->
...
+                      Acc#{PacketId => {Type, Msg, Meta}};
+                  _ ->
+                      Acc
+              end
+          end,
+    ExpiredMsgs = maps:fold(Fun, Inflight, #{}),
+    case maps:size(ExpiredMsgs) of
+        0 -> State;
+        _ ->
+            ExpiredMsgs1 = lists:sort(
```
Comment (zmstone) [inline src/emqtt.erl:1159]: this sort is unnecessary

File: src/emqtt.erl [hunk 4]
```diff
@@ -1073,6 +1141,53 @@ delete_inflight_when_full(Packet, State) ->
...
+delete_expired_inflight_msgs([], State) ->
+    State;
+delete_expired_inflight_msgs(
+  [{PacketId, {Type, Msg, _Meta}} | Msgs],
+  State = #state{inflight = Inflight}
+) ->
+    case Type of
+        publish ->
+            #mqtt_msg{topic = Topic} = Msg,
+            ?LOG(warning, "Message ~p expired, topic: ~ts", [PacketId, Topic], State);
```
Comment (zmstone) [inline src/emqtt.erl:1176]: `info`

File: src/emqtt.erl [hunk 5]
```diff
@@ -1073,6 +1141,53 @@ delete_inflight_when_full(Packet, State) ->
...
+        pubrel ->
+            ?LOG(warning, "Pubrel ~p expired", [PacketId], State)
```
Comment (zmstone) [inline src/emqtt.erl:1178]: `info`

File: src/emqtt.erl [hunk 6]
```diff
@@ -1195,6 +1318,12 @@ deliver(#mqtt_msg{qos = QoS, dup = Dup, retain = Retain, packet_id = PacketId,
     ok = eval_msg_handler(State, publish, Msg),
     State.
 
+now_ts() ->
+    erlang:system_time(millisecond).
+
+timestamp_to_ms(Ts) ->
```
Comment (zmstone) [inline src/emqtt.erl:1324]: why not upgrade Ts from tuple to integer

File: src/emqtt.erl [hunk 7]
```diff
@@ -1380,3 +1509,72 @@ reason_code_name(16#A0) -> maximum_connect_time;
...
+%% Added for 1.2.3.4
+%%
+%% Due to compatibility issues, we can't directly modify the structure of the state to store
+%% these newly added configurations. So we put `max_retry_count` and `message_delivery_timeout`
+%% into the process dictionary temporarily.
+put_max_retry_count(N) when is_integer(N) ->
+    erlang:put(max_retry_count, N).
+
+get_max_retry_count() ->
+    Default = 3,
```
Comment (zmstone) [inline src/emqtt.erl:1525]:
```suggestion
    Default = 0,
```

File: src/emqtt.erl (hunk 6 above)
Comment (HJianBo) [reply src/emqtt.erl:1324]: Mainly to change the timestamp for the flight window from `os:timesamp/0` to `erlang:system_time/1`

File: src/emqtt.erl (hunk 3 above)
Comment (HJianBo) [reply src/emqtt.erl:1159]: The main purpose is to call back the functions in the order in which the messages were sent.

File: src/emqtt.erl (hunk 7 above)
Comment (HJianBo) [reply src/emqtt.erl:1524]: Some test cases require emqtt to do retransmissions, so changing it to 0 here may break compatibility.

**Pull request merged**